### ISSUE TITLE
[REVIEW]: DEVSIM: A TCAD Semiconductor Device Simulator

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3898]
**Submitting author:** @tcaduser (<a href="http://orcid.org/0000-0002-6783-0194">Juan E. Sanchez</a>)
**Repository:** <a href="https://github.com/devsim/devsim.git" target ="_blank">https://github.com/devsim/devsim.git</a>
**Version:** v2.0.1
**Editor:** @lucydot
**Reviewer:** @jgoldfar, @dalonsoa
**Archive:** <a href="https://doi.org/10.5281/zenodo.5908666" target="_blank">10.5281/zenodo.5908666</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/962b5ad390d866433b5efb68b4b68168/status.svg)](https://joss.theoj.org/papers/962b5ad390d866433b5efb68b4b68168)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/962b5ad390d866433b5efb68b4b68168"><img src="https://joss.theoj.org/papers/962b5ad390d866433b5efb68b4b68168/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/962b5ad390d866433b5efb68b4b68168/status.svg)](https://joss.theoj.org/papers/962b5ad390d866433b5efb68b4b68168)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/devsim/devsim.git" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@jgoldfar & @dalonsoa, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @lucydot know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @jgoldfar
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/devsim/devsim.git">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/devsim/devsim.git">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@tcaduser) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @dalonsoa
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/devsim/devsim.git">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/devsim/devsim.git">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@tcaduser) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
PDF failed to compile for issue #3898 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.44 s (1353.8 files/s, 205502.8 lines/s)
----------------------------------------------------------------------------------
Language                        files          blank        comment           code
----------------------------------------------------------------------------------
C++                               195           8695           4117          42382
C/C++ Header                      194           3372           3045           8817
Python                             90           1423           2319           4529
yacc                                3            148             96           1500
Fortran 77                         15              3           2606           1495
CMake                              32            179            134           1012
Bourne Shell                       37            214            252            859
GLSL                                6             70             32            657
lex                                 3            134             50            475
Markdown                            4            127              0            198
Tcl/Tk                              2             28             41            103
Windows Message File                5             27              0             95
YAML                                2             54            127             88
DOS Batch                           1             11              0             40
C Shell                             1              1              0              7
----------------------------------------------------------------------------------
SUM:                              590          14486          12819          62257
----------------------------------------------------------------------------------


Statistical information for the repository '230ca5d69b3653164f198d9b' was
gathered on 2021/11/08.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Juan                             2         65898              0           49.84
Juan E Sanchez                   1            57             48            0.08
Juan E. Sanchez                162         39049          26658           49.70
Juan Sanchez                     3           408             94            0.38

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Juan E. Sanchez           78699          201.5         14.6               13.33
```

**Comment by lucydot** [comment]
@whedon generate pdf from branch publications

**Comment by whedon** [comment]
PDF failed to compile for issue #3898 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by lucydot** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:wave: @jgoldfar, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @dalonsoa, please update us on how your review is going (this is an automated reminder).

**Comment by dalonsoa** [comment]
It's in my ToDo list for this week. Expect an updated by the weekend.

**Comment by dalonsoa** [comment]
I have a problem with testing. 

`devsim` includes a collection of python scripts that run as part of its testing suite and that check the different functionality. However, it uses no standard testing framework for the high level python interface ([`pytest`](https://docs.pytest.org/en/6.2.x/), [`unittest`](https://docs.python.org/3/library/unittest.html)) nor a unit testing framewrok for the low level c++ functinality ([`catch`](https://github.com/catchorg/Catch2)). What there is now is neither exaustive nor sustainable. 

However, requiring to implement any of this is a major and very time consuming undertaking and I am reluctant to block the publication until it is done.

@lucydot , what si your suggestion on this?

**Comment by tcaduser** [comment]
> 

Hello,

The tests are driven via CTest and are run as part of the release system.  For each platform, the regressions are run and new results are updated before each release.

These are the existing regression tests:
https://github.com/devsim/devsim_tests_msys
https://github.com/devsim/devsim_tests_win64
https://github.com/devsim/devsim_tests_macos_gcc_x86_64
https://github.com/devsim/devsim_tests_linux_x86_64

This includes the tests in the ``testing`` and examples in the ``examples`` directory.

**Comment by dalonsoa** [comment]
About the paper:

The description of the **State of the field** is a bit limited. It mentions two open source packages that operate in the same field, but it does not make any comparison with them in terms of features or capabilities. More importantly, it does not compare with the very strong comercial packages (SILVACO, COMSOL). Obviously, one of the benefits of `devsim` is that is is open source and free compared with the $$$$ that the comercial solutions cost, so I think there is a strong case to be made on how this open source tool can leverage the functionality of some of the other, way less accessible, commercial packages.

**Comment by tcaduser** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #3898 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by tcaduser** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: :point_left:

**Comment by tcaduser** [comment]
@dalonsoa 
> About the paper:
> 
> The description of the **State of the field** is a bit limited. It mentions two open source packages that operate in the same field, but it does not make any comparison with them in terms of features or capabilities. More importantly, it does not compare with the very strong comercial packages (SILVACO, COMSOL). Obviously, one of the benefits of `devsim` is that is is open source and free compared with the $$$$ that the comercial solutions cost, so I think there is a strong case to be made on how this open source tool can leverage the functionality of some of the other, way less accessible, commercial packages.

Thank you for your insight.  The TCAD landscape is very broad.  Since I have not used the other OSI-compliant open source offerings, I do not feel it is fair to comment on their capabilities.  Especially in that both Solcore and Charon are subject to rapid change.  Cogenda TCAD (GPL3) has not made any updates to their software in the recent future.

As for the commercial offering, I would find it difficult to compare this tool directly to the offerings of Synopsys, Silvaco, and COMSOL.  From what I can tell, the first two are comprehensive solutions with person decades of experience.  COMSOL is an FEM multiphysics package, that made the extra effort to implement FVM to incorporate Scharfetter Gummel.

I would prefer to not even get started discussing the tools with restrictive closed-source academic licenses.  Stanford's PISCES, Open NESS come to mind.

How much detail would you want for this section?  I have added an additional citation to the paper for a researcher using DEVSIM for DLTS.

**Comment by lucydot** [comment]
RE: testing - @dalonsoa are the queries you raised partly answered by the response from @tcaduser (which seems to relate to a C++ testing unit framework) or are you still hoping for more clarification from editorial point of view?

Ideally the Python interface would have automated testing using pytest or similar, however if there are tests that a user can run manually then this is acceptable. 

@tcaduser It may be that you need to clarify details around testing in the documentation (indicate coverage, what is automated, what is manual, how a user can verify that the installation is correct and the software is working as expected).

**Comment by jgoldfar** [comment]
> 👋 @jgoldfar, please update us on how your review is going (this is an automated reminder).

I am going through the steps in [this listing](https://github.com/openjournals/joss-reviews/issues/3898#issue-1047430015) in this issue, and have found that I cannot toggle the "check boxes" - is this a problem on my end, or is it because I am not assigned to this issue?

**Comment by lucydot** [comment]
Hi @jgoldfar that's weird - you are listed as a reviewer so should have been assigned automatically. I'll try adding you now. Thanks for flagging this up.

**Comment by lucydot** [comment]
Ok, I've added you manually, let's see if it works!

**Comment by tcaduser** [comment]
@jgoldfar, @dalonsoa 

I have merged the ``joss`` branch  with the ``r1.7.0`` branch for the latest versions of INSTALL.md, README.md, . . .  The pdf form of the documentation is in ``doc/devsim.pdf`` of the same branch.

**Comment by tcaduser** [comment]
@jgoldfar, @dalonsoa 

The newest release branch ``r2.0.0`` has been merged into the ``joss`` branch.

**Comment by dalonsoa** [comment]
@tcaduser , I'm happy with the improvements in the installation instructions - great job! - and I've also checked some features and performance. All good on that front. After the clarification above, I am also happy with the testing part - although I strongly recomend opening an issue about implementing a unit testing suite - for both the C++ part and the Python part. But that is work for the future. 

Now, I am still not convinced by the Statements of Purpose in both the paper and the documentation, as well as the State of the Field. I understand the arguments you give [above](https://github.com/openjournals/joss-reviews/issues/3898#issuecomment-980085105), but even that discussion - without actually comparing features, something I agree is very challenging - is extremelly valuable to be written down as part of the documents. This way, people can easily put DEVISIM on the map of TCAD software, which is an enormous field.

Also, the Contributing guidelines. At the very least, I would describe the whole ecosystem of repositories you have in place hosting the source code, the tests for the different platforms, etc, so potential contributors cand find out how things are arranged.

In summary: the tool itself is great, but it is necessary some extra work to clarify where it sits in the big scheme of things and how to support it.

**Comment by tcaduser** [comment]
@dalonsoa please check the updated paper, to see if it meets your expectations.  It adds an extended discussion comparing the simulator to the commercial offerings.  It also sets a list of requirements for the class of TCAD problems it solves.  The paper now explicitly references SYMDIFF, the symbolic differentiation engine that is used by the simulator, and that is also maintained by our project.

On the joss branch
The CONTRIBUTING.md now points back to the README.md, where I have explicitly added the related projects.

**Comment by danielskatz** [comment]
@whedon re-invite @jgoldfar as reviewer

**Comment by whedon** [comment]
The reviewer already has a pending invite.

@jgoldfar please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by lucydot** [comment]
@dalonsoa @jgoldfar @tcaduser please note that I am setting my out of office for the next two weeks over Christmas. I will check in once a week, but my responses might be slower than usual for the next fortnight.

@jgoldfar please let me know if you are still having problems with the tick-boxes, there is an invite link just above this message.

**Comment by dalonsoa** [comment]
@tcaduser , now this looks really, relaly good, both the paper and the new README. Many thanks for taking the time to implementing the changes. 

All good from my side. After merging the JOSS branch, **I recomend accepting the work for its publication in JOSS**.

**Comment by lucydot** [comment]
@dalonsoa thank you for your timely and thoughtful review. 

We are waiting on a second review from @jgoldfar  - could you give us an update on how this is going? I will also contact through email as I know sometimes the github notifications are missed (I am guilty of this anyhow).

**Comment by lucydot** [comment]
@whedon re-invite @jgoldfar as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@jgoldfar please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by jgoldfar** [comment]
Thanks @lucydot - I'm able to check off the tasks for this review now.

@tcaduser - well done on this comprehensive work. I would support a more automated approach to testing as suggested by @dalonsoa for the continued robustness of your project; it looks like you have integration tests as mentioned [in your comment](https://github.com/openjournals/joss-reviews/issues/3898#issuecomment-978052530), though they don't appear to be automated. Unit testing for every component of your application is probably too much to ask, and can create unnecessary fragility - you are probably the best person to decide on right combination of unit tests and automated functional or integration tests.

Recognizing that the technical aspects of the work are more of a journey than a destination, I too recommend **accepting this work for publication in JOSS**.

**Comment by lucydot** [comment]
Hi @jgoldfar that's great - thank you for your comments. If you are happy that the software passes criteria, can I ask you to tick off the appropriate boxes at the top of the issue? There are two unticked. Note that the "automated testing" can be ticked if there are manual steps outlined so that a user can verify that the paper is working as intended - which I believe, from the discussions I've seen, is the case here.

**Comment by jgoldfar** [comment]
Done

**Comment by lucydot** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: :point_left:

**Comment by lucydot** [comment]
@whedon check references

**Comment by lucydot** [comment]
On behalf of the JOSS editorial team, thank you very much @jgoldfar and @dalonsoa for your reviews ✨ 

@tcaduser it looks like we are onto the final lap!

I have a few editorial suggestions for the JOSS article - to improve readability:

- `DEVSIM is a technology computer-aided design (TCAD) software` : --> re-word this to `DEVSIM is technology computer-aided design (TCAD) software` or perhaps `DEVSIM is a technology computer-aided design (TCAD) code` or `DEVSIM is a technology computer-aided design (TCAD) simulator`.
- `This capability is useful for calibration to model existing fabricated devices`: --> I'm not sure I fully understand the meaning here. Perhaps `It can be used to model existing, fabricated devices for calibration purposes`.
- `saving the number of` : --> `reducing the number of`.
- `Software with TCAD simulation`: --> perhaps `Software tools with TCAD simulation` (the former sounds not quite right to me as it is unclear whether it is a single piece of software of multiple pieces of software - my suggestion makes it clear that it is the plural).
- `The number of open source 
offerings, meeting the open source definition (OSD) (Open Source Initiative, 2021), has been
growing (Alonso-Álvarez et al., 2018; Cogenda, 2016; Sandia National Laboratories, 2020)` : --> `The number of open source
offerings, as defined by the open source initiative (Open Source Initiative, 2021), is growing (Alonso-Álvarez et al., 2018; Cogenda, 2016; Sandia National Laboratories, 2020).`
- `To our knowledge, the only OSD simulators, having most of these features, are Solcore,` : --> `To our knowledge the only OSD simulators having most of these features are Solcore,`
- Paragraph beginning`A unique feature of the software is its scripting model interface. It symbolically evaluates expressions and their derivatives.... ` : --> `A unique feature of DEVSIM is its scripting model interface which symbolically evaluates expressions and their derivatives. In this sense, DEVSIM is tailored toward modeling advanced TCAD simulation models. The core engine is maintained as a separate project, SYMDIFF (DEVSIM LLC, 2021b).`

Once you make the necessary changes please let me know here - thanks!

**Comment by tcaduser** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: :point_left:

**Comment by tcaduser** [comment]
@lucydot , thank you for your excellent suggestions.  I have incorporated your changes, with some minor additional edits.

I have also merged in the r2.0.1 branch to the joss branch.  This incorporates additional changes to README.md, CONTRIBUTING.md, INSTALL.md.

**Comment by tcaduser** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: :point_left:

**Comment by tcaduser** [comment]
@lucydot , I have corrected the typo

**Comment by lucydot** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: :point_left:

**Comment by lucydot** [comment]
@tcaduser it is not corrected - I have created a PR: https://github.com/devsim/devsim/pull/82

**Comment by tcaduser** [comment]
@lucydot PR accepted, thanks

**Comment by lucydot** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by lucydot** [comment]
Excellent @tcaduser we are at the final steps now. 

Please can you do the following:
- Make a tagged release of your software  
- Archive this release on Zenodo (or similar)
  - The archive should have exactly the same title and author list as your JOSS paper
 - Report the archive DOI and version number in this thread

**Comment by lucydot** [comment]
@tcaduser -  bumping this message up!

**Comment by tcaduser** [comment]
Sorry for the delay.  I have a DOI number reserved in Zenodo, and I will attempt to make a tagged release in the next 7 days.

Is the doi number of the proof, ``10.21105/joss.03898``, the one that is going to be on the final paper?

**Comment by tcaduser** [comment]
> Excellent @tcaduser we are at the final steps now.
> 
> Please can you do the following:
> 
>* Make a tagged release of your software

Version 2.0.1
https://github.com/devsim/devsim/releases/tag/v2.0.1

>* Archive this release on Zenodo (or similar)

[10.5281/zenodo.5908666](https://zenodo.org/record/5908666)

>   * The archive should have exactly the same title and author list as your JOSS paper
> * Report the archive DOI and version number in this thread

Thank you,

Juan

**Comment by lucydot** [comment]
Thanks @tcaduser that's great 🎈

**Comment by lucydot** [comment]
@whedon set 10.5281/zenodo.5908666 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5908666" target="_blank">10.5281/zenodo.5908666</a> is the archive.

**Comment by whedon** [comment]
OK. v2.0.1 is the version.

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by lucydot** [comment]
I'd like to thank @jgoldfar and @dalonsoa for their reviews 👏

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch publications. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3898 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by lucydot** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: :point_left:

**Comment by lucydot** [comment]
Hello @openjournals/joss-eics I think this paper is ready for the final checks from the EiC team before publication...

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/TED.2021.3094776 is OK
- 10.1088/1361-6463/ac34ad is OK
- 10.1002/pssa.202100408 is OK
- 10.1007/s10825-018-1171-3 is OK
- 10.1002/nme.2579 is OK
- 10.1145/2629697 is OK

MISSING DOIs

- 10.1007/978-3-7091-8752-4 may be a valid DOI for title: Analysis and simulation of semiconductor devices
- 10.15368/theses.2019.60 may be a valid DOI for title: Designing a Simulator for an Electrically-Pumped Organic Laser Diode

INVALID DOIs

- None
```

**Comment by kthyng** [comment]
@tcaduser a few items before we wrap up:
1. version and archive look good ✅ 
2. Can you check out those missing dois, unless you have already checked previously? They aren't always correct.
3. Check the capitalization in your references. For example, Alonso-Alvarez et al has "python" uncapitalized. You can preserve capitalization with {} around characters in your .bib file.

**Comment by tcaduser** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: :point_left:

**Comment by tcaduser** [comment]
Hello @kthyng ,

In addition, The URL for Lucy Whalley in the left panel does not work correctly in macOS Preview.  It probably needs the `https://` prefix

**Comment by kthyng** [comment]
@whedon check references from branch joss

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-7091-8752-4 is OK
- 10.1109/TED.2021.3094776 is OK
- 10.1088/1361-6463/ac34ad is OK
- 10.1002/pssa.202100408 is OK
- 10.15368/theses.2019.60 is OK
- 10.1007/s10825-018-1171-3 is OK
- 10.1002/nme.2579 is OK
- 10.1145/2629697 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by kthyng** [comment]
@openjournals/dev Can you check the links in the produced pdfs? As the author said above, the links for Lucy and for the archive in the left panel of the pdf do not work.

**Comment by danielskatz** [comment]
👋 @openjournals/dev - help please...

**Comment by arfon** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03898/joss.03898/10.21105.joss.03898.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-7091-8752-4 is OK
- 10.1109/TED.2021.3094776 is OK
- 10.1088/1361-6463/ac34ad is OK
- 10.1002/pssa.202100408 is OK
- 10.15368/theses.2019.60 is OK
- 10.1007/s10825-018-1171-3 is OK
- 10.1002/nme.2579 is OK
- 10.1145/2629697 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2966

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2966, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true from branch joss 
```

**Comment by arfon** [comment]
> https://github.com/orgs/openjournals/teams/dev Can you check the links in the produced pdfs? As the author said above, the links for Lucy and for the archive in the left panel of the pdf do not work.

Apologies for the confusion, a couple of issues were going on here:

1. The archive URL only works in the final proofs (they work here https://github.com/openjournals/joss-papers/pull/2966)
2. The URL associated with @lucydot's JOSS profile was missing the `http://` prefix. I've now fixed this.

Both URLs are now working (in the final proofs here:  https://github.com/openjournals/joss-papers/pull/2966)

**Comment by danielskatz** [comment]
@whedon accept deposit=true from branch joss

ok, looks good to me

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1493625071935623169) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2967
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03898](https://doi.org/10.21105/joss.03898)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by danielskatz** [comment]
We're waiting for some backend processing to finish so that this is officially published and the DOI works. We'll add another comment here when this is complete.

**Comment by danielskatz** [comment]
Congratulations to @tcaduser (Juan E. Sanchez)!!

And thanks to @jgoldfar and @dalonsoa for reviewing, and to @lucydot for editing!
We couldn't do this without you!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03898/status.svg)](https://doi.org/10.21105/joss.03898)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03898">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03898/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03898/status.svg
   :target: https://doi.org/10.21105/joss.03898
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03898/status.svg)](https://doi.org/10.21105/joss.03898)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by tcaduser** [comment]
Thanks to everyone for their help.  Can someone give me guidance as to where to request the tagged languages for the software on the publication page?  Here, it is showing as Ruby, Python, Batchfile:
https://joss.theoj.org/papers/10.21105/joss.03898

It should be just Python and C++.

**Comment by arfon** [comment]
> It should be just Python and C++.

I've fixed this for you manually. We simply take what GitHub's Linguist library tells us here.
